jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannel.java

```diff
@@ -64,52 +64,59 @@ public class ServletChannel implements Runnable
 {
     private static final Logger LOG = LoggerFactory.getLogger(ServletChannel.class);
 
+    private final ServletRequestState _state;
     private final AtomicLong _requests = new AtomicLong();
-    private Connector _connector;
-    private Executor _executor;
-    private HttpConfiguration _configuration;
-    private EndPoint _endPoint;
-    private ServletRequestState _state;
+    private final Connector _connector;
+    private final Executor _executor;
+    private final HttpConfiguration _configuration;
+    private final EndPoint _endPoint;
+    private final HttpInput _httpInput;
+    private final Listener _combinedListener;
     private ServletContextHandler.ServletContextApi _servletContextApi;
-    private ServletContextRequest _request;
+    private ServletContextRequest _servletContextRequest;
     private boolean _expects100Continue;
-    private Listener _combinedListener;
     private long _oldIdleTimeout;
     private Callback _callback;
     // Bytes written after interception (e.g. after compression).
     private long _written;
 
+    public ServletChannel(ServletContextHandler servletContextHandler, Request request)
+    {
+        _state = new ServletRequestState(this);
+        _connector = request.getConnectionMetaData().getConnector();
+        _executor = request.getContext();
+        _configuration = request.getConnectionMetaData().getHttpConfiguration();
+        _endPoint = request.getConnectionMetaData().getConnection().getEndPoint();
+        _httpInput = new HttpInput(this);
+        _combinedListener = new Listeners(_connector, servletContextHandler);
+    }
+
     public void setCallback(Callback callback)
     {
         if (_callback != null)
             throw new IllegalStateException();
         _callback = callback;
     }
 
-    public void init(ServletContextRequest request)
+    public void init(ServletContextRequest servletContextRequest)
     {
-        _connector = request.getConnectionMetaData().getConnector();
-        _executor = request.getContext();
-        _configuration = request.getConnectionMetaData().getHttpConfiguration();
-        _endPoint = request.getConnectionMetaData().getConnection().getEndPoint();
-        _state = new ServletRequestState(this); // TODO can this be recycled?
-        _servletContextApi = request.getContext().getServletContext();
-        _request = request;
-        _expects100Continue = request.getHeaders().contains(HttpHeader.EXPECT, HttpHeaderValue.CONTINUE.asString());
-        _combinedListener = new Listeners(request);
-
-        request.getHttpInput().init();
+        _state.recycle();
+        _httpInput.reopen();
+        _servletContextApi = servletContextRequest.getContext().getServletContext();
+        _servletContextRequest = servletContextRequest;
+        _expects100Continue = servletContextRequest.getHeaders().contains(HttpHeader.EXPECT, HttpHeaderValue.CONTINUE.asString());
+        _callback = null;
```

Comment:
`_state.recycle()` and `_callback = null` should be done in `recycle()` method not in the init.